xled/control.py

```diff
@@ -72,6 +72,20 @@ def session(self):
             assert self._session
         return self._session
 
+    def session_get(self, url, retriable = True):
```

Comment:
I'm wondering if this doesn't duplicate `auto_refresh_token` flag of `BaseUrlChallengeResponseAuthSession`? Granted, it isn't set by default in the code.

Or did I misunderstood what you try to achieve here?

Reply:
The `TokenExpiredError` is raised only when calculated token's time to live is expired. In my case, the token was invalidated by another client. Initially I found it while controlling the lights both from mobile app and from python REPL, but the simplest steps to reproduce is described in 6cccbb4e:

```python
ctrl1 = xled.control.ControlInterface(ip)
ctrl2 = xled.control.ControlInterface(ip)

ctrl1.is_on() # => True
ctrl2.is_on() # => True
ctrl1.is_on() # => raises an exception
```
on master, this snippet will throw a JSON decoding error, as 401 response does not have any JSON data in body.

6cccbb4e adds a check for HTTP status, fe8e337 converts the error to TokenExpiredError, and the last two commit handle the case by retrying the failed requests once.

Reply:
I see. I like to handle this kind of issues. On the other hand I'd rather like to have this handling on lower level - by `auth.BaseUrlChallengeResponseAuthSession`. From a quick look it might be possible to hook to responses similarly how `ChallengeResponseAuth` does that. I'm not sure if all the retry handling would fit nicely though.

What do you think?

Reply:
If I understood you, #36 should do the right thing.